coll: autogenerate top-level collective code

hzhou opened the pull request:
## Pull Request Description

The top-level collective code, e.g. `src/mpi/coll/bcast/bcast.c`, are mostly boilerplate, suitable for auto-generation. This will facilitate future modification to collective interfaces, e.g. adding or removing extra parameters.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/ucx

raffenet reviewed (commented):
Is there any way to validate these changes programmatically? Something like simulating a bunch of collective inputs and validating the algorithms that are chosen? I don't know if anything like that was done during the Csel integration, but it would be helpful. There's obviously more than can be reviewed manually here.

hzhou commented:
> 
> 
> Is there any way to validate these changes programmatically? Something like simulating a bunch of collective inputs and validating the algorithms that are chosen? I don't know if anything like that was done during the Csel integration, but it would be helpful. There's obviously more than can be reviewed manually here.

Yes! https://github.com/pmodels/mpich/pull/5368. I was trying to revise the cvar tests so it can test every cvar algorithms. But that PR will need insert many fallback checks, which this PR can make it easy.

Note the Csel selection itself is not touched in this PR.

raffenet commented:
> > Is there any way to validate these changes programmatically? Something like simulating a bunch of collective inputs and validating the algorithms that are chosen? I don't know if anything like that was done during the Csel integration, but it would be helpful. There's obviously more than can be reviewed manually here.
> 
> Yes! #5368. I was trying to revise the cvar tests so it can test every cvar algorithms. But that PR will need insert many fallback checks, which this PR can make it easy.
> 
> Note the Csel selection itself is not touched in this PR.

OK. If you think this needs to be merged, then go ahead. I'd like to see how selection is affected, but I don't think this PR is reviewable in the traditional sense.

hzhou commented:
> OK. If you think this needs to be merged, then go ahead. I'd like to see how selection is affected, but I don't think this PR is reviewable in the traditional sense.

Let me work on #5368 (rebasing on this PR) to verify via tests. Meanwhile, I still believe you can review the python code, which is 680 lines, and maybe also review a few functions in the output. The collective codes are almost all boilerplates, so I believe a few of them can be sufficiently representative.

There are some changes introduced in this PR, mostly are inconsistency in the old code. Notables are:
* added bcast smp cvar
* added gpu buffer swap for persistent op collectives
* added some more COLLECTIVE_FALLBACKS

I can programmatically check and list all these differences if that's what you are looking for.

hzhou commented on src/mpi/coll/mpir_coll_sched_auto.c:
```diff
@@ -0,0 +1,918 @@
+
+#include "mpiimpl.h"
+
+/* This file contains hand coded nonblocking auto selection algorithms.
+ * They are superseded by csel-based selection logic. However, some
+ * non-blocking algorithm (e.g. sched_smp) may need call auto algorithms
+ * on its sub-problems with a pre-allocated schedule. Thus, we are still
+ * defining them here.
+ */
```
> ... after we went through the work of organizing them by collective type not too long ago.

Do you have more information on which work was that?

Ideally we should have a version of csel selection to replace these hand coded algorithms, so I view them as a temporary leftovers.

raffenet replied:
I'm thinking of when we organized collectives by directory. https://github.com/pmodels/mpich/pull/2859/commits/8a48c6bcaa9b2a0ab2060b6331224b1a6f1c4e81

I guess Csel did re-consolidate most of that logic, so I won't strongly object to this change.

hzhou commented on src/mpi/request/mpir_request.c:
```diff
@@ -175,6 +175,12 @@ int MPIR_Request_completion_processing(MPIR_Request * request_ptr, MPI_Status *
                 if (request_ptr->u.persist_coll.real_request != NULL) {
+                    MPII_Coll_req_t *coll = &request_ptr->u.persist_coll.coll;
+                    if (coll->host_recvbuf) {
+                        MPIR_Localcopy(coll->host_sendbuf, coll->count, coll->datatype,
```
fixed

hzhou commented on maint/gen_coll.py:
```diff
@@ -0,0 +1,680 @@
+        (func_intra, algo_list, algo) = (None, None, None)
+        for line in In:
+            if RE.match(r'(\w+-(intra|inter)):', line):
+                func_intra = RE.m.group(1)
```
Fixed. I used `commkind` since the `_` interferes (in my comprehension) with the name patterns that it is composing.

hzhou commented on src/mpi/coll/coll_algorithms.txt:
```diff
@@ -0,0 +1,385 @@
+barrier-intra:
+    dissemination
+    smp
+        restrictions: has_subcomm
```
Documentation added at the top of this file.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most
